### PR TITLE
Output run time per stream

### LINKED ISSUE #7965: Python CDK: output runtime per stream
## Tell us about the problem you're trying to solve
When debugging long running syncs, it is often helpful to know how long it took to run particular streams. 

Currently we have no explicit tracking which says how long it took to run particular streams. 

## Describe the solution you’d like
1. As each stream completes, output how long it took to sync the stream
1. Once the sync is successful, output how long it took to run each stream
It would be helpful if this was formatted in a human readable way and alphabetically sorted e.g: 
```
Stream runtimes: 
A: 1h 2m 34s
B: 0h 2m 0s
C: 0h 15m 12s
```
If the connector fails halfway through, then output as much of this information as possible regardless (output a partial table)

### PR DESCRIPTION
**Pull request by eliziario**
## What
Fixes #7965 
## How
An EventTimer class is introduced along with a context manager. Due to the possibility that streams could be run in parallell in the future, care has been taken to ensure EventTimer is thread safe

## Recommended reading order
1. `abstract_source.py`
2. `event_timing.py`

## 🚨 User Impact 🚨
Connectors need to rebuilt for change to be reflected 

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>

#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [Y ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

File: airbyte-integrations/bases/base-python/base_python/cdk/abstract_source.py [hunk 1]
```diff
@@ -78,15 +79,25 @@ def read(
...
-                yield from self._read_stream(
-                    logger=logger, stream_instance=stream_instance, configured_stream=configured_stream, connector_state=connector_state
-                )
-            except Exception as e:
-                logger.exception(f"Encountered an exception while reading stream {self.name}")
-                raise e
+                for configured_stream in catalog.streams:
+                    try:
+                        stream_instance = stream_instances[configured_stream.stream.name]
+                        timer.start_event(configured_stream.stream.name)
```
Comment (avida) [inline airbyte-integrations/bases/base-python/base_python/cdk/abstract_source.py:87]: What is exception while _read_stream occure? Why you dont using context manager here as well?

File: airbyte-integrations/bases/base-python/base_python/cdk/utils/event_timing.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        self.count = 0
+        EventTimer.thread_local.stack = []
+        self.stack = EventTimer.thread_local.stack
+        self.lock = Lock()
+
+    def start_event(self, name):
+        """
+        Start a new event and push it to the stack.
+        """
+        self.lock.acquire()
```
Comment (avida) [inline airbyte-integrations/bases/base-python/base_python/cdk/utils/event_timing.py:38]: why you need thread synchronization? airbyte cdk is not using and planning to use multi threading.

File: airbyte-integrations/bases/base-python/base_python/cdk/utils/event_timing.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+class EventTimer:
+    """Simple nanosecond resolution event timer for debugging, initially intended to be used to record streams execution
+    time for a source.
+       As it is likely that streams could be read in parallel, we use a thread local storage to store the
+       stack of events and thus keep track of the current event.
+       Event nesting follows a LIFO pattern, so finish will apply to the last started event.
+    """
+
+    thread_local = local()
```
Comment (keu) [inline airbyte-integrations/bases/base-python/base_python/cdk/utils/event_timing.py:24]: I agree with @avida here, thread safety wasn't a requirement.

File: airbyte-integrations/bases/base-python/base_python/cdk/utils/event_timing.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        self.lock = Lock()
+
+    def start_event(self, name):
+        """
+        Start a new event and push it to the stack.
+        """
+        self.lock.acquire()
+        self.events[name] = Event(name=name)
+        self.count += 1
+        self.lock.release()
```
Comment (keu) [inline airbyte-integrations/bases/base-python/base_python/cdk/utils/event_timing.py:41]: also, why not use context manager here?

File: airbyte-integrations/bases/base-python/base_python/cdk/abstract_source.py (hunk 1 above)
Comment (eliziario) [reply airbyte-integrations/bases/base-python/base_python/cdk/abstract_source.py]: fixed

File: airbyte-integrations/bases/base-python/base_python/cdk/utils/event_timing.py (hunk 3 above)
Comment (eliziario) [reply airbyte-integrations/bases/base-python/base_python/cdk/utils/event_timing.py]: Removed this

File: airbyte-integrations/bases/base-python/base_python/cdk/utils/event_timing.py (hunk 4 above)
Comment (eliziario) [reply airbyte-integrations/bases/base-python/base_python/cdk/utils/event_timing.py]: removed

File: airbyte-integrations/bases/base-python/base_python/cdk/utils/event_timing.py (hunk 2 above)
Comment (eliziario) [reply airbyte-integrations/bases/base-python/base_python/cdk/utils/event_timing.py]: Removed it

**Review by avida — APPROVED** [review]

**Pull request merged**